test/infrastructure/inmemory/Dockerfile

```diff
@@ -0,0 +1,74 @@
+# syntax=docker/dockerfile:1.4
+
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Build the manager binary
+# Run this with docker build --build-arg builder_image=<golang:x.y.z>
+ARG builder_image
+
+# Build architecture
+ARG ARCH
+
+# Ignore Hadolint rule "Always tag the version of an image explicitly."
+# It's an invalid finding since the image is explicitly set in the Makefile.
+# https://github.com/hadolint/hadolint/wiki/DL3006
+# hadolint ignore=DL3006
+FROM ${builder_image} as builder
+
+# Run this with docker build --build-arg goproxy=$(go env GOPROXY) to override the goproxy
+ARG goproxy=https://proxy.golang.org
+ENV GOPROXY=$goproxy
+
+# Gets additional CAPIM dependencies
+WORKDIR /tmp
+
+WORKDIR /workspace
+COPY go.mod go.mod
+COPY go.sum go.sum
+
+# Essentially, change directories into the test go module
+WORKDIR /workspace/test
+# Copy the Go Modules manifests
+COPY test/go.mod go.mod
+COPY test/go.sum go.sum
+
+# Cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN --mount=type=cache,target=/go/pkg/mod \
+    go mod download
+
+# This needs to build with the entire Cluster API context
+WORKDIR /workspace
+# Copy in cluster-api (which includes the test/infrastructure/inmemory subdirectory)
+COPY . .
+
+# Essentially, change directories into CAPIM
+WORKDIR /workspace/test/infrastructure/inmemory
+
+# Build
+ARG ARCH
+
+# Build the CAPIM manager using the compiler cache folder
+RUN --mount=type=cache,target=/root/.cache/go-build \
```

Comment:
Let's please reduce the diff to the top-level Dockerfile as far as possible, e.g.:
* support lflags arg
* use nonroot image
* probably more

Reply:
WDTY would it be possible to use the same Dockerfile but have a some args for the differences? (can be a follow-up issue, we need to fix the current issue ASAP in my opinion)

Reply:
I changed the Dockerfile to be as close to the root level Dockerfile as possible.